src/channels/pages/ChannelDetailsPage/ChannelDetailsPage.tsx

```diff
@@ -20,38 +37,101 @@ export interface ChannelDetailsPageProps {
   saveButtonBarState: ConfirmButtonTransitionState;
   onBack?: () => void;
   onDelete?: () => void;
-  onSubmit?: (data: FormData) => void;
+  onSubmit: (data: FormData) => void;
   updateChannelStatus?: () => void;
 }
 
-const initialData: FormData = {
-  currencyCode: "",
-  name: "",
-  slug: ""
-};
-
 export const ChannelDetailsPage: React.FC<ChannelDetailsPageProps> = ({
   channel,
   currencyCodes,
   disabled,
   disabledStatus,
+  onSubmit,
   errors,
   onBack,
-  onSubmit,
   onDelete,
   saveButtonBarState,
   updateChannelStatus
 }) => {
-  const [selectedCurrencyCode, setSelectedCurrencyCode] = React.useState("");
+  const [selectedCurrencyCode, setSelectedCurrencyCode] = useState("");
+
+  const [shippingZonesToDisplay, setShippingZonesToDisplay] = useStateFromProps<
+    Channel_channel_shippingZones[]
+  >(channel?.shippingZones || []);
+
+  const initialData: FormData = {
+    currencyCode: "",
+    name: "",
+    slug: "",
+    shippingZonesIdsToAdd: [],
+    shippingZonesIdsToRemove: [],
+    ...channel
+  };
+
+  const {
+    loadMore: fetchMoreShippingZones,
+    search: searchShippingZones,
+    result: searchShippingZonesResult
+  } = useShippingZonesSearch({
+    variables: DEFAULT_INITIAL_SEARCH_DATA
+  });
+
+  const getFilteredShippingZonesChoices = () =>
+    getParsedSearchData(searchShippingZonesResult).filter(
+      ({ id: searchedZoneId }) =>
+        !shippingZonesToDisplay.map(({ id }) => id).includes(searchedZoneId)
+    );
 
   return (
-    <Form onSubmit={onSubmit} initial={channel || initialData}>
-      {({ change, data, hasChanged, submit }) => {
+    <Form onSubmit={onSubmit} initial={initialData}>
+      {({ change, data, hasChanged, submit, triggerChange, set }) => {
         const handleCurrencyCodeSelect = createSingleAutocompleteSelectHandler(
           change,
           setSelectedCurrencyCode,
           currencyCodes
         );
+
+        const addShippingZone = (zoneId: string) => {
+          set({
```

Comment:
It seems like `set()` should be triggering `setChanged`. There's only 1 other place I could find that uses `set` (`PluginsDetailsPage.tsx` line 102) and it's also used in combination with `triggerChange`. I suggest adding that logic to the `useForm` hook and using a second optional argument to disable the trigger:

```
// useForm.tsx

  function set(newData: Partial<T>, triggerChange = true) {
    setData(data => ({
      ...data,
      ...newData
    }));

  if (triggerChange) { onChange(true) }
  }
```

Reply:
Ideally this entire hook form should be using `useEffect` with 1-level deep comparison to trigger `onChange` automatically instead of calling it directly inside the change methods